chore(Rln-relay): Tutorial on how to run rln-chat2 in on-chain mode (#1044)
status-im/nwaku: docs/tutorial/onchain-rln-relay-chat2.md

In the on-chain/dynamic group management, the state of the group members i.e., their identity commitment keys is moderated via a membership smart contract deployed on the Goerli network which is one of the Ethereum testnets.
Members can be dynamically added to the group and the group size can grow up to 2^20 members.
This is in contrast to the prior test scenarios in which the rln group was static and the set of members' keys was hardcoded and fixed.

## Prerequisites 
In this tutorial, you will need 1) an account with at least `0.001` ethers on the Goerli testnet and 2) a hosted node on the Goerli testnet. 
In case you are not familiar with either of these two steps, you may use the following tutorial on the [prerequisites of running on-chain spam-protected chat2](./pre-requisites-of-running-on-chain-spam-protected-chat2.md).
Note that the required `0.001` ethers correspond to the registration fee, 
however, you still need to have more funds in your account to cover the cost of the transaction gas fee.

## Overview
Figure 1 provides an overview of the interaction of the chat2 clients with the test fleets and the membership contract. 

> At a high level,  when a chat2 client is spun up with Waku-RLN-Relay mounted in on-chain mode, it creates rln credentials (i.e., an identity key and an identity commitment key) and 

Author: Thanks, to avoid confusion removed the plural form. 
 Please see 8aa8bb2

In the following test setting, the chat2 clients are to be connected to the Waku test fleets as their first hop. 
The test fleets will act as routers and are also set to run Waku-RLN-Relay over the same pubsub topic and content topic as chat2 clients i.e., the default pubsub topic of `/waku/2/default-waku/proto` and the content topic of `/toy-chat/2/luzhou/proto`. 
Spam messages published on the said combination of topics will be caught by the test fleet nodes and will not be routed.
Note that spam protection does not rely on the presence of the test fleets.
In fact, all the chat2 clients are also capable of catching and dropping spam messages if they receive any.
You can test it by connecting two chat2 clients (running Waku-RLN-Relay) directly to each others and see they can spot each others' spam activities.

 ![](./imgs/rln-relay-chat2-overview.png)
 Figure 1.

# Set up
## Build chat2
First, build chat2 with the RLN flag set to true.

```
make chat2 RLN=true
```

## Set up a chat2 client

Run the following command to set up your chat2 client. 

```

> ./build/chat2  --fleet:test --content-topic:/toy-chat/2/luzhou/proto --rln-relay:true --rln-relay-dynamic:true --eth-mem-contract-address:0x5DE1Fb10345Ef1647629Df30e6C397297Da09A1d  --eth-account-address:xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx --eth-account-privatekey:xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx  --eth-client-address:xxxx  --ports-shift=1  

Reviewer: Thanks @staheri14 for PR. Contract address is incorrect. Needs to be `0x4252105670fE33D2947e8EaD304969849E64f2a6`

Author: done in 0feab5efc211dfd2570144c50e7282bba4d80237

```

In this command
- the `--fleet:test` indicates that the chat2 app gets connected to the test fleets.
- the `toy-chat/2/luzhou/proto` passed to the `content-topic` option indicates the content topic on which the chat2 application is going to run.
- the `rln-relay` flag is set to `true` to enable the Waku-RLN-Relay protocol for spam protection.
- the `--rln-relay-dynamic` flag is set to `true`  to enable the on-chain mode of  Waku-RLN-Relay protocol with dynamic group management.
- the `--eth-mem-contract-address` option gets the address of the membership contract.

>   The current address of the contract is `0x5DE1Fb10345Ef1647629Df30e6C397297Da09A1d`.

Reviewer: Same. Contract address is incorrect. Needs to be `0x4252105670fE33D2947e8EaD304969849E64f2a6`

Author: done in 0feab5efc211dfd2570144c50e7282bba4d80237

You will see a different value than `165886530` on your screen. 
If two messages sent by the same chat2 client happen to have the same RLN epoch value, then one of them will be detected as spam and won't be routed (by test fleets in this test setting).
At the time of this tutorial, the epoch duration is set to `10` seconds.
You can inspect the current epoch value by checking the following [constant variable](https://github.com/status-im/nim-waku/blob/21cac6d491a6d995a7a8ba84c85fecc7817b3d8b/waku/v2/protocol/waku_rln_relay/waku_rln_relay_types.nim#L119) in the nim-waku codebase.
Thus, if you send two messages less than `10` seconds apart, they are likely to get the same `rln epoch` values.

After sending a chat message, you may experience some delay before the next chat prompt appears. 
The reason is that under the hood a zero-knowledge proof is being generated and attached to your message.

Try to spam the network by violating the message rate limit i.e.,
sending more than one message per epoch. 
Your messages will be routed via test fleets that are running in spam-protected mode over the same content topic i.e., `/toy-chat/2/luzhou/proto` as your chat client.

> Your samp activity will be detected by them and your message will not reach the rest of the chat clients.

Reviewer 2: ```suggestion
Your spam activity will be detected by them and your message will not reach the rest of the chat clients.
```

> You can check this by running a second chat user and verifying that spam messages are filtered at the test fleets and not displayed. 

Reviewer 2: ```suggestion
You can check this by running a second chat user to verify that spam messages are not displayed because are filtered by the test fleets.
```

A sample test scenario is illustrated in the [Sample test output section](#sample-test-output).

Once you are done with the test, make sure you close all the chat2 clients by typing the `/exit` command.
```
>> /exit
quitting...
```

## How to reuse rln credentials

You may resue your old rln credentials using `rln-relay-membership-index`, `rln-relay-id` and `rln-relay-id-commitment` options. 
For instance, if the previously generated credentials are
```
your membership index is: 63
your rln identity key is: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
your rln identity commitment key is: 6c6598126ba10d1b70100893b76d7f8d7343eeb8f5ecfd48371b421c5aa6f012
```
Then, the execution command will look like this (inspect the last three config options):
```

> ./build/chat2  --fleet:test --content-topic:/toy-chat/2/luzhou/proto --rln-relay:true --rln-relay-dynamic:true --eth-mem-contract-address:0x5DE1Fb10345Ef1647629Df30e6C397297Da09A1d  --eth-account-address:your_eth_account --eth-account-privatekey:your_eth_private_key  --eth-client-address:your_goerli_node  --ports-shift=1  --rln-relay-membership-index:63 --rln-relay-id:xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx --rln-relay-id-commitment:6c6598126ba10d1b70100893b76d7f8d7343eeb8f5ecfd48371b421c5aa6f012

Reviewer 2: why not pass just the --rln-relay-id ? If we need the commitment we then do a Poseidon hash over it. The commitment is used for something specific? In fact, to generate proofs we only need the id-key and the tree index (locally we already have the commitment as part of the tree leaves).

Author: Obviously id commitment is recomputable from id key, however, having it passed by the user makes them maintain the commitment key part which will enable them to look up their key state on the membership smart contract. If they only keep the id key part, then they will need to recompute the hash (hence knowing the hash parameters) to look up their key state on the membership smart contract. 

As for the usability of the id commitment: atm, it is not being used as a critical data piece in the code, but we may need it later to announce slashed keys in a p2p manner or other cases.

> ./build/chat2  --fleet:test --content-topic:/toy-chat/2/luzhou/proto --rln-relay:true --rln-relay-dynamic:true --eth-mem-contract-address:0x5DE1Fb10345Ef1647629Df30e6C397297Da09A1d  --eth-account-address:your_eth_account --eth-account-privatekey:your_eth_private_key  --eth-client-address:your_goerli_node  --ports-shift=1  --rln-relay-membership-index:63 --rln-relay-id:xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx --rln-relay-id-commitment:6c6598126ba10d1b70100893b76d7f8d7343eeb8f5ecfd48371b421c5aa6f012

Reviewer: Contract address is incorrect. Needs to be `0x4252105670fE33D2947e8EaD304969849E64f2a6`

Author: Thanks, fixed in 0feab5efc211dfd2570144c50e7282bba4d80237
